### PR TITLE
Treat blocked/ready labels as states

### PR DESCRIPTION
**Pull request by discorick**
Merges https://github.com/huboard/huboard-web/pull/327  (the card design branch.)

Blocked/Ready labels now act readily as states on HuBoard. If the label is missing <s>it is created</s> the health check fails. 

closes https://github.com/huboard/huboard/issues/399

### CODE REVIEW

**Comment by dahlbyk** [comment]
> For discussion sake you can cause strange behavior by:
> 
> - Assigning both a blocked/ready label to an issue, this will enable both states at once
> 
> The algorithm isn't overly bright, it will just pick the first state label and run with it. Personally, I am comfortable with this - when wrestling with outside influences (like label changes from GitHub) theres only so much we can do to guard against non-sensical states.

I think it reasonable for us to assign reasonable priority to our labels. If something is Blocked, it is explicitly not Ready. Once un-Blocked, it's reasonable to immediately fall back into a Ready state. My $.02.



> If the label is missing it is created.

I'm unsure about this, since many boards already have some sort of "impediment" label that they might want to just rename for use by HuBoard (ideally we'd make the Blocked label configurable - `huboard.yml` strikes again!).

What if we added a Block/Ready label checker to the health check? "[Click Here] to create, or rename your existing labels to..."

Nice-to-have: option in health check to apply the labels to currently blocked/ready-without-label cards.

**Comment by discorick** [comment]
> I think it reasonable for us to assign reasonable priority to our labels. If something is Blocked, it is explicitly not Ready. Once un-Blocked, it's reasonable to immediately fall back into a Ready state. My $.02.

I suppose before each label update on the client we can make sure only "one or the other" before we make any PUT's. The desired state will be the first in the in the array.

> I'm unsure about this, since many boards already have some sort of "impediment" label that they might want to just rename for use by HuBoard (ideally we'd make the Blocked label configurable - huboard.yml strikes again!).

Until we improve our UX around how we communicate features, the health check route will likely just push this feature into the dark along with other less obvious HB features. I propose revisit this once we have yml config rolled out.

**Comment by dahlbyk** [comment]
OK, so...thoughts on how to communicate either:

1) You don't have permission to create labels, so your Ready/Block state couldn't be set with a label, or
2) We just created a label for you, so you might want to flip into GitHub, delete it and rename your existing Impediment label to match HB's expectation.

------

> Until we improve our UX around how we communicate features, the health check route will likely just push this feature into the dark along with other less obvious HB features.

We should add analytics to identify how many boards are "unhealthy", and how often the health page is used.

**Comment by discorick** [comment]
> 1) You don't have permission to create labels, so your Ready/Block state couldn't be set with a label

Being able to label/unlabel an issue means you are a collaborator, implicitly you also can create them. 

> 2) We just created a label for you, so you might want to flip into GitHub, delete it and rename your existing Impediment label to match HB's expectation.

Flash message: "HuBoard just created the blocked label for you"

**Comment by dahlbyk** [comment]
> Being able to label/unlabel an issue means you are a collaborator, implicitly you also can create them.

Touché

> Flash message: "HuBoard just created the blocked label for you"

That works. Sticky? Can we add a link to FAQ to suggest renaming an existing label, if they have one?

**Comment by discorick** [comment]
> That works. Sticky? Can we add a link to FAQ to suggest renaming an existing label, if they have one?

Seems good to me.

> Once un-Blocked, it's reasonable to immediately fall back into a Ready state.

Won't be too bad to handle in real-time on the client. The effort required to sort out how to handle an issue with both blocked and ready labels from GitHub seems more costly than its worth. Like I said before, it's just nonsensical to begin with.

**Comment by dahlbyk** [comment]
> Like I said before, it's just nonsensical to begin with.

It is, but deterministic handling doesn't seem prohibitively expensive. The same is true of our column labels - are we nondeterministic, or is there a rule (e.g. highest number wins)?

**Comment by discorick** [comment]
> It is, but deterministic handling doesn't seem prohibitively expensive. The same is true of our column labels - are we nondeterministic, or is there a rule (e.g. highest number wins)?

Well, It's already deterministically choosing blocked state if both labels are there, I just don't think we should remove the counterpart label automatically via the API.

**Comment by discorick** [comment]
@dahlbyk, starting to warm up to your point of view about automatically creating labels. I don't love the UX of it. We can start by just treating labels that are 'blocked' or 'ready' as states, then later figure out a good way to treat them as a health check ? (Or maybe in this PR)

**Comment by discorick** [comment]
As per @dahlbyk , should the label picker deselect 'blocked' labels if 'ready' is selected or vice versa ?

File: ember-app/app/models/new/issue.js [hunk 1]
```diff
@@ -142,24 +169,18 @@ var Issue = Model.extend({
...
   },
   stateLabel: function(){
-    var state_label = this.get('other_labels').map((label)=> {
-        return label.name.toLowerCase();
-      }).find((name)=>{
-        return name.toLowerCase() === 'blocked' || name.toLowerCase() === 'ready';
-    });
-    return state_label || "";
+    return this.get('other_labels').map((label)=>{return label.name.toLowerCase()}).find((name)=>{
+      return name === 'blocked' || name === 'ready';
```
Comment (dahlbyk) [inline ember-app/app/models/new/issue.js]: Could this just be `this.get('other_labels').find(this.isStateLabel)`?

File: ember-app/app/models/new/issue.js [hunk 2]
```diff
@@ -142,24 +169,18 @@ var Issue = Model.extend({
...
-  customState: Ember.computed("data._data.custom_state", "data.other_labels.[]", {
+  customState: Ember.computed("data._data.custom_state", "data.other_labels.[]", "stateLabel", {
     get:function(){
       var state = this.get("stateLabel");
       if(state){ return state; }
       return this.get("_data.custom_state");
     },
     set: function (key, value) {
-      var previousState = this.get("stateLabel") || this.get("_data.custom_state");
+      var previousState =  this.get("stateLabel") || this.get("_data.custom_state");
```
Comment (dahlbyk) [inline ember-app/app/models/new/issue.js]: :trollface:

File: lib/health_checking/health_checks/board/issue_blocked_label_check.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+module HealthChecking
+  module HealthChecks
+    module Board
+      class IssueBlockedLabelCheck
+        include HealthCheck
+        name 'issue_blocked_label_check'
+        weight :warning
+        authorization :collaborator
+        passed "Issue blocked label is present"
+        failed "Issue blocked label is not present"
```
Comment (dahlbyk) [inline lib/health_checking/health_checks/board/issue_blocked_label_check.rb]: Maybe `"Blocked" label is not present`? "Issue" seems redundant, and maybe more explicitly mentioning the expected label name would nudge people toward renaming their own? Or we could add some sort of tooltip to give a help bubble with more explanation.

**Comment by dahlbyk** [comment]
Other than clarifying the health check for those who might already have equivalent-ish labels, I think we're pretty good here.

> should the label picker deselect 'blocked' labels if 'ready' is selected or vice versa ?

I feel pretty strongly that using the label picker directly should behave like a pure label picker. If you happen to pick both Ready and Blocked (or pick Blocked while Ready is already selected, or whatever), then our state UX should respect the label with the highest precedence (Blocked) but keep labels as-selected.

**Comment by dahlbyk** [comment]
Two observations from just a bit more testing:

- Moving a blocked/ready card between columns is changing state in the description, but not removing the label.
- Not new to this PR, but it seems like Blocked state should only be removed if the card is moved to the right on the board. I could see a blocked card in Next Release moving back into Backlog while still blocked.

File: ember-app/app/models/new/issue.js (hunk 1 above)
Comment (discorick) [inline ember-app/app/models/new/issue.js]: Labels don't have their own their own intelligent models, so theres not really an efficient way other than checking every single label on its creation and setting the bool (as opposed to a computed property).

File: ember-app/app/models/new/issue.js (hunk 2 above)
Comment (discorick) [inline ember-app/app/models/new/issue.js]: 😲

File: lib/health_checking/health_checks/board/issue_blocked_label_check.rb (hunk 3 above)
Comment (discorick) [inline lib/health_checking/health_checks/board/issue_blocked_label_check.rb]: Yeah, that copy is better, issue is redundant. 

> Or we could add some sort of tooltip to give a help bubble with more explanation.

I think for our UX 'funnel' phase we should identify things like this and determine a good way to provide intuitive info about HuBoard on the fly

**Comment by discorick** [comment]
> Moving a blocked/ready card between columns is changing state in the description, but not removing the label.

Shoot! That _should_ be working, I will look into it. 

> Not new to this PR, but it seems like Blocked state should only be removed if the card is moved to the right on the board. I could see a blocked card in Next Release moving back into Backlog while still blocked.

Hmm, thats probably true in many cases (where the issue remains blocked). Do we want to tackle this here though?

File: ember-app/app/models/new/issue.js (hunk 1 above)
Comment (dahlbyk) [inline ember-app/app/models/new/issue.js]: Oh, I mean passing `this.isStateLabel` as a function as the `filter` callback. `isStateLabel` doesn't use `this` at all, it's a proper function (input :point_right: output).

**Comment by dahlbyk** [comment]
> Do we want to tackle this here though?

Probably not, just thought I'd mention it while it was top of mind.

**Comment by discorick** [comment]
> Moving a blocked/ready card between columns is changing state in the description, but not removing the label.

Found the problem, sneaky edge case good catch.

**Comment by dahlbyk** [comment]
:tada:

**Pull request merged**